Mathlib/Geometry/Euclidean/Incenter.lean

```diff
@@ -0,0 +1,622 @@
+
+noncomputable section
+
+namespace Affine
+
+namespace Simplex
+
+variable {n : ℕ} [NeZero n] (s : Simplex ℝ P n)
+
+open scoped Classical in
+/-- The unnormalized weights of the vertices in an affine combination that gives an excenter with
+signs determined by the given set of indices (for the empty set, this is the incenter; for a
+singleton set, this is the excenter opposite a vertex).  An excenter with those signs exists if
+and only if the sum of these weights is nonzero (so the normalized weights sum to 1). -/
+def excenterWeightsUnnorm (signs : Set (Fin (n + 1))) (i : Fin (n + 1)) : ℝ :=
+  (if i ∈ signs then -1 else 1) * (dist (s.points i)
+    (orthogonalProjection (affineSpan ℝ (Set.range (s.faceOpposite i).points)) (s.points i)))⁻¹
+
+/-- Whether an excenter exists with a given choice of signs. -/
+def ExcenterExists (signs : Set (Fin (n + 1))) : Prop :=
+  ∑ i, s.excenterWeightsUnnorm signs i ≠ 0
+
+/-- The normalized weights of the vertices in an affine combination that gives an excenter with
+signs determined by the given set of indices.  An excenter with those signs exists if and only if
+the sum of these weights is 1. -/
+def excenterWeights (signs : Set (Fin (n + 1))) : Fin (n + 1) → ℝ :=
+  (∑ i, s.excenterWeightsUnnorm signs i)⁻¹ • s.excenterWeightsUnnorm signs
+
+@[simp] lemma excenterWeightsUnnorm_compl (signs : Set (Fin (n + 1))) :
+    s.excenterWeightsUnnorm signsᶜ = -s.excenterWeightsUnnorm signs := by
+  ext i
+  by_cases h : i ∈ signs <;> simp [excenterWeightsUnnorm, h]
+
+@[simp] lemma excenterWeights_compl (signs : Set (Fin (n + 1))) :
+    s.excenterWeights signsᶜ = s.excenterWeights signs := by
+  simp [excenterWeights, inv_neg]
+
+@[simp] lemma excenterExists_compl {signs : Set (Fin (n + 1))} :
+    s.ExcenterExists signsᶜ ↔ s.ExcenterExists signs := by
+  simp [ExcenterExists]
+
+lemma sum_excenterWeights (signs : Set (Fin (n + 1))) [Decidable (s.ExcenterExists signs)] :
+    ∑ i, s.excenterWeights signs i = if s.ExcenterExists signs then 1 else 0 := by
+  simp_rw [ExcenterExists, excenterWeights]
+  split_ifs with h
+  · simp [← Finset.mul_sum, h]
+  · simp only [ne_eq, not_not] at h
+    simp [h]
+
+@[simp] lemma sum_excenterWeights_eq_one_iff {signs : Set (Fin (n + 1))} :
+    ∑ i, s.excenterWeights signs i = 1 ↔ s.ExcenterExists signs := by
+  classical
+  simp [sum_excenterWeights]
+
+alias ⟨_, ExcenterExists.sum_excenterWeights_eq_one⟩ := sum_excenterWeights_eq_one_iff
+
+lemma sum_excenterWeightsUnnorm_empty_pos : 0 < ∑ i, s.excenterWeightsUnnorm ∅ i := by
+  refine Finset.sum_pos ?_ (by simp)
+  rintro i -
+  simp [excenterWeightsUnnorm]
+
+/-- The existence of the incenter, expressed in terms of `ExcenterExists`. -/
+lemma excenterExists_empty : s.ExcenterExists ∅ :=
+  s.sum_excenterWeightsUnnorm_empty_pos.ne'
+
+lemma sum_inv_dist_orthogonalProjection_faceOpposite_sq_smul_vsub_eq_zero :
+    ∑ i, (dist (s.points i)
+      (orthogonalProjection (affineSpan ℝ (Set.range (s.faceOpposite i).points))
```

Comment:
Do you think we should have a name for `affineSpan _ (Set.range (Simplex.points s)))`? `Simplex.span`?

Reply:
We do in fact have `orthogonalProjectionSpan` in `Circumcenter.lean`, which would also serve the purpose of making these expressions shorter (though maybe there is a case for having both definitions). If we want to make more use of `orthogonalProjectionSpan`, I should definitely start by factoring orthogonal projection definitions and API out (from a mixture of `Basic.lean` and `Circumcenter.lean`) into a separate file - which should be done at a time when there aren't any outstanding open PRs such as #23712 and #23715 to that material.

Reply:
Those PRs are now merged; do you think I makes sense to do this now, or not until after this PR?

Reply:
I think we should treat these as independent changes.

I've filed #24389 with the split of orthogonal projections to a separate file, which would be a prerequisite for making more use of `orthogonalProjectionSpan`.

Reply:
I don't quite follow why the split needs to happen first; can we just import the file here in the short term so that we can make the 16 occurences of `orthogonalProjection (affineSpan ℝ (Set.range (s.faceOpposite` much shorter? We can play the game of pruning the imports later (as this is a leaf file anyway), but I think we should have the nice spelling available ASAP.

Reply:
Oh, I thought you meant an issue when you said "filed", and were proposing leaving it for another time. I've approved the PR :)

Reply:
Now that split is in, I've changed this PR to use `orthogonalProjectionSpan` whenever it was previously using an expanded version thereof.